pkg/controller/migmigration/stage.go

```diff
@@ -0,0 +1,415 @@
+package migmigration
+
+import (
+	"context"
+	"reflect"
+	"strconv"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+
+	migapi "github.com/konveyor/mig-controller/pkg/apis/migration/v1alpha1"
+	migref "github.com/konveyor/mig-controller/pkg/reference"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// StagePod - wrapper for stage pod, allowing to compare  two stage pods for equality
+type StagePod struct {
+	corev1.Pod
+}
+
+// StagePodList - a list of stage pods, with built-in stage pod deduplication
+type StagePodList []StagePod
+
+// BuildStagePods - creates a list of stage pods from a list of pods
+func BuildStagePods(list *corev1.PodList) StagePodList {
+	stagePods := StagePodList{}
+	for _, pod := range list.Items {
+		stagePods.merge(buildStagePodFromPod(migref.ObjectKey(&pod), pod.Labels, &pod))
+	}
+	return stagePods
+}
+
+func (p StagePod) equals(pod StagePod) bool {
+	for _, volume := range p.Spec.Volumes {
+		found := false
+		for _, targetVolume := range pod.Spec.Volumes {
+			if reflect.DeepEqual(volume.VolumeSource, targetVolume.VolumeSource) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return false
+		}
+	}
+	return true
+}
+
+func (l *StagePodList) contains(pod StagePod) bool {
+	for _, srcPod := range *l {
+		if pod.equals(srcPod) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (l *StagePodList) merge(list ...StagePod) {
+	for _, pod := range list {
+		if !l.contains(pod) {
+			*l = append(*l, pod)
+		}
+	}
+}
+
+func (l *StagePodList) create(client k8sclient.Client) (int, error) {
+	existingPods := StagePodList{}
+	if len(*l) > 0 {
+		existingPods.list(client, (*l)[0].Labels)
+	}
+	counter := 0
+	for _, stagePod := range *l {
+		if existingPods.contains(stagePod) {
+			continue
+		}
+		err := client.Create(context.TODO(), &stagePod.Pod)
+		if err != nil && !errors.IsAlreadyExists(err) {
+			return 0, err
+		}
+		counter++
+	}
+
+	return counter + len(existingPods), nil
+}
+
+func (l *StagePodList) list(client k8sclient.Client, labels map[string]string) error {
+	podList := corev1.PodList{}
+	options := k8sclient.MatchingLabels(labels)
+	err := client.List(context.TODO(), options, &podList)
+	if err != nil {
+		return err
+	}
+	for _, pod := range podList.Items {
+		l.merge(buildStagePodFromPod(migref.ObjectKey(&pod), pod.Labels, &pod))
+	}
+
+	return nil
+}
+
+func (t *Task) ensureStagePodsFromOrphanedPVCs() error {
+	existingStagePods := StagePodList{}
+	stagePods := StagePodList{}
+	client, err := t.getSourceClient()
+	if err != nil {
+		log.Trace(err)
+		return err
+	}
+
+	err = existingStagePods.list(client, t.stagePodLabels())
+	if err != nil {
+		log.Trace(err)
+		return nil
+	}
+
+	pvcMounted := func(list StagePodList, claimName string) bool {
```

Comment:
It will, but I think this is better option then allow to select different behaviour for 2 PVCs in different namespaces. No matter what, if the first case ignores `skip` option for the PVC, it will be selected for copy there. During the stage pod creation phase, only one stage pod will be created, due to the fact that both stage pods will have equal mount fields. So `copy` + `skip` or `copy` + `copy` will have the same outcome here.

Reply:
But if pvcMounted() returns true for a volume that's orphaned, it means the stage pod won't be created, and the volume won't be copied. PVCs are namespaced, which means the same named pvcs in two different namespaces are completely different PVCs. They point to different PVs.

Reply:
Good catch :D